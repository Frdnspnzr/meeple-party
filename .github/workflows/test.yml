name: Test
on:
  pull_request:
    branches:
      - "*"
  workflow_dispatch:
    inputs:
      update-snapshots:
        description: "Update snapshots?"
        type: boolean
permissions:
  contents: read
  pages: write
  id-token: write
jobs:
  build:
    timeout-minutes: 10
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: lts/*
          cache: "npm"
      - name: Install dependencies
        run: npm install
      - name: Generate database bindings
        run: npm run "prisma generate"
      - name: Build app
        run: npm run build
      - name: Save build
        uses: actions/upload-artifact@v4
        with:
          name: build
          path: .next

  e2e:
    timeout-minutes: 60
    runs-on: ubuntu-latest
    needs: build
    services:
      postgres:
        image: postgres
        env:
          POSTGRES_PASSWORD: postgres
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: lts/*
          cache: "npm"
      - name: Install dependencies
        run: npm install
      - name: Install Playwright browsers
        run: npx playwright install --with-deps
      - name: Load build
        uses: actions/download-artifact@v4
        with:
          name: build
          path: .next
      - name: Set up snapshot cache
        id: snapshots_cache
        uses: actions/cache@v4
        with:
          key: ${{ runner.os }}-snapshots-${{github.ref}}
          restore-keys: ${{ runner.os }}-snapshots-refs/heads/main
          path: ./tests/e2e/snapshots/**
      - name: Update snapshots
        if: ${{steps.snapshots_cache.outputs.cache-hit != 'true' || inputs.update-snapshots == 'true'}}
        run: npx playwright test --update-snapshots
      - name: Generate database bindings
        run: npm run "prisma generate"
      - name: Generate database tables
        run: npm run "prisma deploy"
        env:
          DATABASE_URL: postgresql://postgres:postgres@localhost:5432/test
      - name: Run Playwright tests
        run: npx playwright test --max-failures=10
        env:
          DATABASE_URL: postgresql://postgres:postgres@localhost:5432/test
          NEXT_PUBLIC_SECRET: Testrunner
          NEXTAUTH_URL: http://127.0.0.1:3000
      - uses: actions/upload-artifact@v4
        if: always()
        with:
          name: playwright-report
          path: playwright-report/

  unittest:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: lts/*
          cache: "npm"
      - name: Install dependencies
        run: npm install
      - name: Generate database bindings
        run: npm run "prisma generate"
      - name: Run the tests
        run: npm test
      - name: Upload coverage reports to Codecov
        uses: codecov/codecov-action@v3
        env:
          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}

  publish-e2e:
    runs-on: ubuntu-latest
    needs: [e2e, build]
    if: ${{ !cancelled() && (needs.build.result=='success') }}
    steps:
      - name: Configure GitHub Pages
        uses: actions/configure-pages@v3
      - name: Load report
        uses: actions/download-artifact@v4
        with:
          name: playwright-report
          path: playwright-report/
      - name: Upload Pages Artifact
        uses: actions/upload-pages-artifact@v2
        with:
          path: playwright-report/
      - name: Deploy to GitHub Pages
        uses: actions/deploy-pages@v2
