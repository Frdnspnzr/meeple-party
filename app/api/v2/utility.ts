import { Relationship, RelationshipType } from "@/datatypes/relationship";
import {
  UserPreferences,
  UserProfile,
  defaultUserPreferences,
} from "@/datatypes/userProfile";
import {
  Prisma,
  Relationship as PrismaRelationship,
  User,
} from "@prisma/client";

export type FullPrismaRelationship = PrismaRelationship & {
  sender: User;
  recipient: User;
};

export function convertToUserProfile(
  user: User,
  friend: boolean = false
): UserProfile {
  return {
    id: user.id,
    name: user.name,
    role: user.role,
    image: user.image,
    realName: allowRealName(user, friend) ? user.realName : null,
    about: user.about,
    place: allowPlace(user, friend) ? user.place : null,
    bggName: allowBggName(user, friend) ? user.bggName : null,
  };
}

export function normalizeRelationship(
  prismaRelationship: FullPrismaRelationship,
  userId: string
): Relationship {
  return {
    profile: getProfile(prismaRelationship, userId),
    type: getRelationshipType(prismaRelationship, userId),
    lastUpdate: prismaRelationship.updatedAt,
  };
}

function allowRealName(user: User, friend: boolean): boolean {
  if (friend) {
    return true;
  }
  const preferences = getPreferences(user);
  return preferences.showRealNameInProfile;
}

function allowPlace(user: User, friend: boolean): boolean {
  if (friend) {
    return true;
  }
  const preferences = getPreferences(user);
  return preferences.showPlaceInProfile;
}

function allowBggName(user: User, friend: boolean): boolean {
  return friend;
}

function getPreferences(user: User): UserPreferences {
  return {
    ...defaultUserPreferences,
    ...(user.preferences as Prisma.JsonObject),
  };
}

function getRelationshipType(
  relationship: FullPrismaRelationship,
  userId: string
): RelationshipType {
  if (relationship.type === "FRIENDSHIP") {
    return RelationshipType.FRIENDSHIP;
  } else if (relationship.type === "FRIEND_REQUEST") {
    if (relationship.sender.id === userId) {
      return RelationshipType.FRIEND_REQUEST_SENT;
    } else {
      return RelationshipType.FRIEND_REQUEST_RECEIVED;
    }
  }
  throw new Error("Error determining relationship type");
}

function getProfile(
  relationship: FullPrismaRelationship,
  userId: string
): UserProfile {
  const profile =
    relationship.recipientId === userId
      ? relationship.sender
      : relationship.recipient;
  return convertToUserProfile(profile, relationship.type === "FRIENDSHIP");
}

/**
 * Reads the authentication token for cronjobs from the environment variable `CRON_AUTH_TOKEN`. Throws when that variable is not set.
 * @returns Authentication token.
 * @throws Throws when no token is configured.
 */
export function getCronAuthToken() {
  if (!process.env.CRON_AUTH_TOKEN) {
    throw new Error("No cron auth token configured for current environment");
  }
  return process.env.CRON_AUTH_TOKEN;
}
