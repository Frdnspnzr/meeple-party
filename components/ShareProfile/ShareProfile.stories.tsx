import type { Meta, StoryObj } from "@storybook/react";

import ShareProfile from "./ShareProfile";

const meta = {
  title: "Share Profile",
  component: ShareProfile,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
} satisfies Meta<typeof ShareProfile>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    profileId: "clnhj5o0n00002ct679u4mwd7",
  },
};

export const NoNativeSharing: Story = {
  args: {
    profileId: "clnhj5o0n00002ct679u4mwd7",
    disableNative: true,
  },
};
