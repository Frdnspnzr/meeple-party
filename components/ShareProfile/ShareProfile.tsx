"use client";

import { Button, Group } from "@mantine/core";
import { useCallback } from "react";

export interface ShareProfileProps {
  profileId: string;
}

export default function ShareProfile({ profileId }: ShareProfileProps) {
  const share = useCallback(() => {
    if (navigator.canShare()) {
      navigator.share({
        url: `${process.env.BASE_URL}/app/profile/${profileId}`,
      });
    }
  }, [profileId]);
  return (
    <Group>
      <Button variant="filled" onClick={share}>
        Button
      </Button>
    </Group>
  );
}
