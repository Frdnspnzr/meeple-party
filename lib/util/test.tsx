import {
  generateArray,
  generateBoolean,
  generateDate,
  generateGame,
} from "@/utility/test";
import { Game as ServerGame } from "@prisma/client";
import { Collection } from "../datatypes/client/collection";
import { Game } from "../datatypes/client/game";
import { UserId } from "../datatypes/client/userProfile";

export function ObjectData({ object }: { object: object }) {
  return (
    <>
      {Object.entries(object).map(([key, value]) => (
        <div data-testid={key} key={key}>
          {JSON.stringify(value)}
        </div>
      ))}
    </>
  );
}

export function generateCollection(userId: UserId): {
  collection: Collection;
  games: Game[];
} {
  const games = generateArray(generateGame);
  const randomGames = () => pickRandom(games).map((g) => g.id);
  const collection = {
    user: userId,
    own: randomGames(),
    wantToPlay: randomGames(),
    wishlist: randomGames(),
  };

  return {
    games,
    collection,
  };
}

export function convertClientGame(game: Game): ServerGame {
  return {
    ...game,
    thumbnail: game.thumbnail || null,
    image: game.image || null,
    BGGRank: game.BGGRank || null,
    BGGRating: game.BGGRating || null,
    updatedAt: generateDate(),
  };
}

export function pickRandom<T>(arr: T[]): T[] {
  return arr.filter(generateBoolean);
}
