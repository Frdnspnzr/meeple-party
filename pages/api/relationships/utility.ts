
import { Relationship, RelationshipType } from "@/datatypes/relationship";
import { PrivateUserProfile, PublicUserProfile } from "@/datatypes/userProfile";
import { prisma } from "@/db";
import { Session } from "@auth0/nextjs-auth0";
import { UserProfile as Auth0Profile } from "@auth0/nextjs-auth0/client";
import {
  Relationship as PrismaRelationship,
  UserProfile,
} from "@prisma/client";

export type FullPrismaRelationship = PrismaRelationship & {
  sender: UserProfile;
  recipient: UserProfile;
};

export function normalizeRelationship(
  prismaRelationship: FullPrismaRelationship,
  userId: number
): Relationship {
  return {
    profile: getProfile(prismaRelationship, userId),
    type: getRelationshipType(prismaRelationship, userId),
    lastUpdate: prismaRelationship.updatedAt,
  };
}

export function getRelationshipType(
  relationship: FullPrismaRelationship,
  userId: number
): RelationshipType {
  if (relationship.type === "FRIENDSHIP") {
    return RelationshipType.FRIENDSHIP;
  } else if (relationship.type === "FRIEND_REQUEST") {
    if (relationship.sender.id === userId) {
      return RelationshipType.FRIEND_REQUEST_SENT;
    } else {
      return RelationshipType.FRIEND_REQUEST_RECEIVED;
    }
  }
  throw new Error("Error determining relationship type");
}

export function getProfile(
  relationship: FullPrismaRelationship,
  userId: number
): PrivateUserProfile | PublicUserProfile {
  const profile =
    relationship.recipientId === userId
      ? relationship.sender
      : relationship.recipient;
  if (
    relationship.type === "FRIENDSHIP" ||
    isFriendRequestReveiced(relationship, userId)
  ) {
    return convertToPrivateProfile(profile);
  } else {
    return convertToPublicProfile(profile);
  }
}

export function convertToPrivateProfile(profile: UserProfile): PrivateUserProfile {
  return {
    ...convertToPublicProfile(profile),
    realName: profile.realName,
  };
}

export function convertToPublicProfile(profile: UserProfile): PublicUserProfile {
  return {
    id: profile.id,
    name: profile.name,
    picture: profile.picture,
  };
}

export function isFriendRequestReveiced(
  relationship: FullPrismaRelationship,
  userId: number
): boolean {
  return (
    relationship.type === "FRIEND_REQUEST" &&
    relationship.recipientId === userId
  );
}

export async function getUserProfile(session: Session): Promise<UserProfile> {
  const user = session.user as Auth0Profile;
  if (!user.email) {
    throw new Error("User from Auth0 does not have an E-Mail adress");
  }
  const userProfile = await prisma.userProfile.findUnique({
    where: { email: user.email as string },
  });

  if (!userProfile) {
    throw new Error("No user profile corresponding to Auth0 user found.");
  }

  return userProfile;
}