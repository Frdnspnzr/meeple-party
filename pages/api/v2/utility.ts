import { authOptions } from "@/pages/api/auth/[...nextauth]";
import { User } from "@prisma/client";
import { NextApiRequest, NextApiResponse } from "next";
import { getServerSession } from "next-auth/next";
import { ApiError } from "./types";

type ApiHandler = (req: NextApiRequest, res: NextApiResponse) => Promise<void>;
type ApiHandlerWithUser = (
  req: NextApiRequest,
  res: NextApiResponse,
  user: User
) => Promise<void>;

export function withUser(handler: ApiHandlerWithUser): ApiHandler {
  return async (req, res) => {
    const session = await getServerSession(req, res, authOptions);
    if (!session || !session.user) {
      throw Error("Unauthorized");
    }
    return await handler(req, res, session.user);
  };
}

export function withErrorHandling(handler: ApiHandler): ApiHandler {
  return async (req, res) => {
    try {
      return await handler(req, res);
    } catch (reason) {
      res.status(401).send({ reason } as ApiError);
    }
  };
}
