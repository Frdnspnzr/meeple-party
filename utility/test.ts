import { FullPrismaRelationship } from "@/app/api/v2/utility";
import { defaultUserPreferences } from "@/datatypes/userProfile";
import { RelationshipType, Role, User } from "@prisma/client";

test("Foo", () => {});

export function generateString(length = 32) {
  let result = "";
  const characters =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  while (result.length < length) {
    result += characters.charAt(Math.floor(Math.random() * characters.length));
  }
  return result;
}

export function generateArray<T>(generator: () => T, length = 10): T[] {
  const ret: T[] = [];
  for (let i = 0; i < length; i++) {
    ret.push(generator());
  }
  return ret;
}

export function generatePrismaUser(): User {
  return {
    id: generateString(25),
    name: generateString(25),
    email: generateString(25),
    profileComplete: true,
    role: Role.USER,
    emailVerified: new Date(),
    about: generateString(1000),
    bggName: generateString(15),
    image: null,
    place: generateString(15),
    realName: generateString(25),
    preferences: defaultUserPreferences,
  };
}

export function generateFullPrismaRelationship(
  type: RelationshipType = RelationshipType.FRIENDSHIP,
  userA?: User,
  userB?: User
): FullPrismaRelationship {
  if (!userA) {
    userA = generatePrismaUser();
  }
  if (!userB) {
    userB = generatePrismaUser();
  }

  return {
    senderId: userA.id,
    recipientId: userB.id,
    type: type,
    sender: userA,
    recipient: userB,
    updatedAt: new Date(),
    cratedAt: new Date(),
  };
}
