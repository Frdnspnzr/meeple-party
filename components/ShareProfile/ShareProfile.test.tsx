import { render } from "@/utility/test";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ShareProfile from "./ShareProfile";

jest.mock("@/i18n/client");

describe("Share Profile", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });
  it("matches snapshot", () => {
    const { container } = render(<ShareProfile profileId="abc" />);
    expect(container).toMatchSnapshot();
  });
  it("matches snapshot without native sharing", () => {
    const { container } = render(
      <ShareProfile profileId="abc" disableNative />
    );
    expect(container).toMatchSnapshot();
  });
  it("matches snapshot without native sharing after clicking share button", async () => {
    const user = userEvent.setup();
    const { container } = render(
      <ShareProfile profileId="abc" disableNative />
    );
    const button = screen.getByRole("button");
    await user.click(button);
    expect(container).toMatchSnapshot();
  });
});
