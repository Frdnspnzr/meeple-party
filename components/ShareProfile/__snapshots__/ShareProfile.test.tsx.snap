// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Share Profile matches snapshot 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m-4081bf90 mantine-Group-root"
    style="--group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;"
  >
    <button
      class="mantine-focus-auto mantine-active m-77c9d27d mantine-Button-root m-87cf2631 mantine-UnstyledButton-root"
      data-variant="filled"
      data-with-left-section="true"
      role="button"
      style="--button-bg: var(--mantine-color-blue-filled); --button-hover: var(--mantine-color-blue-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
      type="button"
    >
      <span
        class="m-80f1301b mantine-Button-inner"
      >
        <span
          class="m-a74036a mantine-Button-section"
          data-position="left"
        >
          <svg
            class="tabler-icon tabler-icon-share-2"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 9h-1a2 2 0 0 0 -2 2v8a2 2 0 0 0 2 2h10a2 2 0 0 0 2 -2v-8a2 2 0 0 0 -2 -2h-1"
            />
            <path
              d="M12 14v-11"
            />
            <path
              d="M9 6l3 -3l3 3"
            />
          </svg>
        </span>
        <span
          class="m-811560b9 mantine-Button-label"
        >
          Share
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`Share Profile matches snapshot without native sharing 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m-4081bf90 mantine-Group-root"
    style="--group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;"
  >
    <button
      class="mantine-focus-auto mantine-active m-77c9d27d mantine-Button-root m-87cf2631 mantine-UnstyledButton-root"
      data-variant="filled"
      data-with-left-section="true"
      role="button"
      style="--button-bg: var(--mantine-color-blue-filled); --button-hover: var(--mantine-color-blue-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
      type="button"
    >
      <span
        class="m-80f1301b mantine-Button-inner"
      >
        <span
          class="m-a74036a mantine-Button-section"
          data-position="left"
        >
          <svg
            class="tabler-icon tabler-icon-share-2"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 9h-1a2 2 0 0 0 -2 2v8a2 2 0 0 0 2 2h10a2 2 0 0 0 2 -2v-8a2 2 0 0 0 -2 -2h-1"
            />
            <path
              d="M12 14v-11"
            />
            <path
              d="M9 6l3 -3l3 3"
            />
          </svg>
        </span>
        <span
          class="m-811560b9 mantine-Button-label"
        >
          Share
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`Share Profile matches snapshot without native sharing after clicking share button 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m-4081bf90 mantine-Group-root"
    style="--group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;"
  >
    <div
      class="m-46b77525 mantine-InputWrapper-root mantine-TextInput-root"
      style="transition-property: transform, opacity; transition-duration: 250ms; transition-timing-function: ease; transform-origin: right; opacity: 1; transform: translateX(0);"
    >
      <div
        class="m-6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
        data-variant="default"
        data-with-right-section="true"
      >
        <input
          aria-invalid="false"
          class="m-8fb7ebe7 mantine-Input-input mantine-TextInput-input"
          data-variant="default"
          id="mantine-n5u7tf7v7"
          style="width: 35ch;"
          value="undefined/app/profile/abc"
        />
        <div
          class="m-82577fc2 mantine-Input-section mantine-TextInput-section"
          data-position="right"
        >
          <button
            class="mantine-focus-auto mantine-active m-8d3f4000 mantine-ActionIcon-root m-87cf2631 mantine-UnstyledButton-root"
            data-size="lg"
            data-variant="subtle"
            style="--ai-size: var(--ai-size-lg); --ai-radius: var(--mantine-radius-xs); --ai-bg: transparent; --ai-hover: var(--mantine-color-blue-light-hover); --ai-color: var(--mantine-color-blue-light-color); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
            type="button"
          >
            <svg
              class="tabler-icon tabler-icon-copy"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 7m0 2.667a2.667 2.667 0 0 1 2.667 -2.667h8.666a2.667 2.667 0 0 1 2.667 2.667v8.666a2.667 2.667 0 0 1 -2.667 2.667h-8.666a2.667 2.667 0 0 1 -2.667 -2.667z"
              />
              <path
                d="M4.012 16.737a2.005 2.005 0 0 1 -1.012 -1.737v-10c0 -1.1 .9 -2 2 -2h10c.75 0 1.158 .385 1.5 1"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
